Fix: #24023 - Ensure different language in translation split view

eremannisto opened the pull request:
## Scope

What's changed:

- Modified language selection logic in split-view to prevent duplicate languages
- Improved TypeScript typing for null safety
- Added better fallback logic for language selection

**Before, would open with the same language:**
<img width="836" alt="before" src="https://github.com/user-attachments/assets/11e8c3e5-9be8-4f39-a11c-bdb7c04ff4c0">

**All languages added to the `languages` collection:**
<img width="888" alt="languages" src="https://github.com/user-attachments/assets/58b6beee-bdba-4c70-9d74-70538f8015a6">

**Default langauge is set to English:**
<img width="888" alt="english" src="https://github.com/user-attachments/assets/4119328c-cc84-4685-97fa-56b47887688e">

**Default language is set to Finnish:**
<img width="888" alt="finnish" src="https://github.com/user-attachments/assets/a745a81f-62f0-4560-943c-635e73248d6c">

**No default language selected:**
<img width="888" alt="none" src="https://github.com/user-attachments/assets/36832959-c012-4832-a064-67682b480d3c">

**Only one language set in the `languages` collection:**
<img width="888" alt="one" src="https://github.com/user-attachments/assets/e69c5ff6-1dcc-46d5-8901-27cd05a6a438">


## Potential Risks / Drawbacks

- None identified - changes only affect the UI behavior of translation split-view
- Maintains existing fallback behavior if only one language is available

## Review Notes / Questions

- The fix prioritizes showing different languages in split-view while respecting user/default language preferences
- The changes include TypeScript improvements for better type safety

---

Fixes [#24023](https://github.com/directus/directus/issues/24023)

Commit:
fix: ensure split-view opens with different languages

When opening translation split-view, the second view now correctly displays a different language instead of duplicating the first view's language.

Also improves TypeScript typing for null safety.

Commit:
Revert "fix: ensure split-view opens with different languages"

This reverts commit 688ebed7fb3e377e6f2ea136e22ef222e0200ec3.

Commit:
fix: ensure split-view opens with different languages

When opening translation split-view, the second view now correctly displays a different language instead of duplicating the first view's language.

Also improves TypeScript typing for null safety.

eremannisto committed:
Update contributors.yml

Commit:
style: format translations interface with prettier

Commit:
Merge branch 'fix-24023/ensure-different-language-in-translation-split-view' of https://github.com/eremannisto/directus into fix-24023/ensure-different-language-in-translation-split-view

eremannisto commented:
Hi! First time contributing to Directus here. Please let me know if any adjustments are needed to align with the project's PR guidelines.

eremannisto committed:
Merge branch 'main' into fix-24023/ensure-different-language-in-translation-split-view

AlexGaillard reviewed (dismissed):
Thanks for the PR @eremannisto 👋 

Great work finding the source of the problem and also the fix. I've included some thoughts about how we can maybe simplify this a little bit and maybe be a bit more performant.

I think we can get this wrapped up pretty quickly 🚀

AlexGaillard commented on app/src/interfaces/translations/translations.vue:
```diff
@@ -242,8 +245,10 @@ function useLanguages() {
 				const defaultLocale = userLanguage.value ? defaultLanguage.value : null;
-				const lang = languages.value.find((lang) => lang[pkField] === defaultLocale) || languages.value[0];
-				secondLang.value = lang?.[pkField];
+				const alternativeLang = languages.value.find((lang) => lang[pkField] !== firstLang.value);
+				const defaultLang = languages.value.find((lang) => lang[pkField] === defaultLocale);
+
+				secondLang.value = (alternativeLang || defaultLang || languages.value[0])?.[pkField];
```
I think this comment got doubled when github conked out on me during review 🙃

Commit:
refactor: optimize language selection logic

Simplify and improve performance of language selection in split-view by removing redundant checks and minimizing array operations. Remove unnecessary type assertion.

eremannisto commented:
Thanks @AlexGaillard for the detailed code review and optimization suggestions! I made the suggested changes and tested it, and seems to work perfeclty. :)

rijkvanzanten commented:
Thanks @eremannisto! And welcome to the repo 🙂 I'll let @AlexGaillard smack that merge button. LGTM! 🚀

AlexGaillard committed:
Merge branch 'main' into fix-24023/ensure-different-language-in-translation-split-view

rijkvanzanten committed:
Add missing newline in contributors
